### ISSUE TITLE
Saving Entries in a Job run by Horizon messes up the Stache

### ISSUE THREAD

**Issue by Buffalom** [statamic/cms#2993]
## Bug Description

I have created a collection containing ordered entries. A laravel job which I execute from the terminal is queued to redis and worked by horizon.
This job gets all published entries of the collection I created, does some changes and executes the save method on them. I have tried this in a fresh project where I don't even do any changes and only execute the save method so it really does not do anything else. When certain changes are made without restarting horizon this job messes up the stache and for instance makes a newly created entry disappear without the file of that entry actually being deleted.

## How to Reproduce

1. Start horizon.
2. Reorder the entries in the collection.
3. Execute the job from the terminal.
4. Refresh the collection in the browser. --> Everything is as expected.
5. Create and publish a new entry in the collection.
6. Execute the job from the terminal.
7. Refresh the collection in the browser. --> The previously created entry is gone. The collection shows exactly as it was in step 4.
8. Refresh the stache. (php please stache:refresh) --> The previously created entry appears again just as it was after step 5.
9. Execute the job from the terminal. --> The previously created entry disappears again, just like it did after step 6.
10. Refresh the stache.
11. Restart horizon. --> Now it is all as it is expected after all the changes. The order is changed and the created entry is visible. If the steps are redone from the beginning now, the previously created entry will not disappear anymore but newly created entries will disappear again.

## Extra Detail

I will gladly show you the problem in a call while sharing my screen and/or answer questions via Slack or Discord or anything like that if that could be of help to you.

### Workaround

This can be worked around if you set horizon balancing to `auto` and the minimum processes to `0`.

## Environment

**Statamic version:** 3.0.31 Solo

**PHP version:** 7.4.12

No addons installed

**Install method (choose one):**
- Fresh install from `statamic/statamic`

**Comment by jelleroorda** [comment]
@damcclean Not sure why you marked this as custom?

**Comment by jasonvarga** [comment]
He's created a custom job that does stuff.

Let's see the code.

**Comment by Buffalom** [comment]
https://github.com/teamnovu/stache-mess-up-test

Username: foo@example.com
Password: foobar

[This](https://github.com/teamnovu/stache-mess-up-test/blob/master/app/Console/Commands/SaveAllEntries.php#L41) is the command you can use to try it out. If you dispatch the job to the queue  `works-here` it will work using the workaround I previously explained.

**Comment by damcclean** [comment]
Thanks for providing a demo repository! However, I don't seem to be able to replicate your issue 🤔 

After I create my entry, I run the command which dispatches to Horizon and your custom job processes. I then clear the Stache using `php artisan stache:clear` and I visit the collection in the CP. My test entry is still there.

**Comment by Buffalom** [comment]
Thanks for the video, and sorry for my late reply.
I've noticed that I forgot to mention I only set up the horizon supervisors for the production environment meaning the `APP_ENV` has to be set to `production` for it to work (or rather not work). However, it looks like you had realized this yourself given your jobs were processed anyway.

I just tried it again on the latest Statamic version (3.0.43 Solo) and the issue persists for me. I also created a [video](https://www.icloud.com/iclouddrive/0J4LglB-9bMfK5SfWMAMT9u0Q#Statamic_Stache_Mess_Up) for you in case that helps.
In your video the `Statamic\StaticCaching\Invalidate` job was processed after my custom job which isn't happening in my video. Might there be something I forgot to set up for this to work properly?
Or is there any other change you made that could have solved the issue for you?

**Comment by sheggi** [comment]
With the reproducing steps and the sample code I've been able to reproduce the behaviour as described.
I used the .env.example with this changes
```
APP_ENV=production
APP_KEY=base64:wMEMhz6pQ8cWhOcqQnlsYFlfaM6I3VdJ+AY5jqikYjQ=
QUEUE_CONNECTION=redis
```

As soon as I added a horizon worker for the `default` queue I couldn't reproduce this behaviour. As I understand the Statamic\StaticCaching\Invalidate job does the job of `php please stache:refresh`.
But this doesn't solve the problem of the horizon-worker somehow keeping a cached stated as of the time when the worker was started.

**Comment by sheggi** [comment]
In addition, once step 7 reached this bug messes with the ordering too.
When I tried different actions including reordering of the entries, the entry appeared again after a browser refresh or appeared at different positions before and after stache:clear or saving in the worker.

**Comment by Buffalom** [comment]
@sheggi Did you encounter an exception? In #3493 they received an exception but I cannot recall getting one.

**Comment by sheggi** [comment]
When I've tested this example, there haven't any errors logged.

I think #3493 may have the same cause. For example when the queue worker has an old stache state, which was not updated on creating a new entry.

**Comment by sheggi** [comment]
@duncanmcclean I guess you didn't run the command before adding a new entry. This step causes the lock-in of the stache state to the running worker. The `php please stache:refresh` has no effect to the state in the worker.

To update or unlock the stache state within the worker these steps are required.
20. `php please stache:refresh`
21. kill or restart worker
22. run command to lock-in the new state

If the command, the SaveAllEntries **job** in particular, is run between step 20. and 21. the old state won't be updated, even the worker has been killed. This also proves that the global stache state could be reset from a worker with an old state.

**Comment by jasonvarga** [comment]
Can you try commenting out these lines and see if the problem continues?

https://github.com/statamic/cms/blob/31e7573c3e71e4477eb0817d906cc39a5c84bc51/src/Stache/Stores/Store.php#L286-L288

**Comment by sheggi** [comment]
> Can you try commenting out these lines and see if the problem continues?
> 
> https://github.com/statamic/cms/blob/31e7573c3e71e4477eb0817d906cc39a5c84bc51/src/Stache/Stores/Store.php#L286-L288
> 
> You'll need to restart horizon.

Without these lines, the problem can't be reproduced any more. 👍🏼

**Comment by avdb** [comment]
Any news on this issue? When will this be solved? Having also an issue that the stache is no representation of de content files of a collection.
Is there a way to get Entry:query() Without using the stache?

**Comment by avdb** [comment]
> Can you try commenting out these lines and see if the problem continues?
> 
> https://github.com/statamic/cms/blob/31e7573c3e71e4477eb0817d906cc39a5c84bc51/src/Stache/Stores/Store.php#L286-L288
> 
> You'll need to restart horizon.

@jasonvarga This works for me as well, but with this we need to overwrite a core class. On production we also do composer install, so we don't submit the vendor files.
Any other solution

**Comment by jelleroorda** [comment]
@avdb https://rias.be/blog/using-composer-patches

**Comment by jesseleite** [comment]
This should be fixed in #6726 now! Anyone able to give it a whirl, maybe via [composer patch](https://rias.be/blog/using-composer-patches)?

**Comment by Buffalom** [comment]
> This should be fixed in #6726 now! Anyone able to give it a whirl, maybe via [composer patch](https://rias.be/blog/using-composer-patches)?

@jesseleite 
I've tried it briefly a few days ago and it seems to be working. Thanks for your fix.
